### ISSUE TITLE
ccd Problem

### ISSUE THREAD

**Issue by jtenner** [schteppe/p2.js#186]
Attached are gifs that show the problem WITH and WITHOUT ccdSpeedThreshold set.

Simply uncommenting the ccdSpeedThreshold variable yields really weird results.

```javascript
body = new Body({
  position: bullet.position,
  angle: bullet.angle,
  //ccdSpeedThreshold: this.initialBulletSpeed * 0.4,
  angularDamping: 1,
  mass: 5,
  damping: 0.01,
  velocity: [
    this.initialBulletSpeed * Math.cos(bullet.angle),
    this.initialBulletSpeed * Math.sin(bullet.angle)
  ]
});
shape = new Circle({
  radius: 4,
  collisionGroup: CollisionGroups.BULLETS,
  collisionMask: CollisionGroups.ENEMIES | CollisionGroups.TERRAIN
});
body.addShape(shape);
world.addBody(body);
```

No ccd:

![No ccd](http://zippy.gfycat.com/WelcomeHardtofindHamadryad.gif)

ccd turned on:
![ccd on](http://zippy.gfycat.com/CaringSmallGraywolf.gif)

ccd on with arc bullet positions:
![ccd on with arc bullet positions](http://zippy.gfycat.com/ObviousRashDowitcher.gif)

The visual shows the small round bullets being fired from the player position fired at an invisible plane wall. I was hoping that I could use ccd for bullets because they are visually entering the boxes, sometimes for two whole frames. It's visually unappealing, but not game breaking.

Any workarounds/help?

Thank you for your excellent JS library.

**Comment by schteppe** [comment]
And you didn't add that invisible plane? Super strange. I need a simple repro... Any idea what it can be?

**Comment by jtenner** [comment]
@schteppe  there is an invisible plane there.

I'm looking at the code that runs, and it moves the bullet to the point of collision, and does nothing with it. (the binary algorithm is really neat.)

Is this intended behavior for now?  It appears to do nothing else with the bullet after it detects a collision.

**Comment by schteppe** [comment]
I see. Cool with web worker too!

That is exactly the intended algorithm. It should move it to the point of collision, leaving a small overlap. Since the body still has velocity, the collision should resolve normally..

**Comment by jtenner** [comment]
http://requirebin.com/embed?gist=be6956ccf80f40761ff1

Here you go. I've replicated the problem :).

Thank you for your effort. Please let me know if there is anything else I can do to help!  

Aside: I'll make sure to share the web-worker framework I use with you when it is completed.

**Comment by jtenner** [comment]
```javascript
if(this.emitImpactEvent && this.has('impact')){
  var ev = this.impactEvent;
  for(var i=0; i!==np.contactEquations.length; i++){
    var eq = np.contactEquations[i];
    if(eq.firstImpact){
      ev.bodyA = eq.bodyA;
      ev.bodyB = eq.bodyB;
      ev.shapeA = eq.shapeA;
      ev.shapeB = eq.shapeB;
      ev.contactEquation = eq;
      this.emit(ev);
    }
  }
}
```

it appears that `this.has('impact');` returns false, implying there are no collisions.

**Comment by schteppe** [comment]
I don't think that part is causing the error... It's something in the CCD algorithm itself. Looking at it a bit now.

**Comment by schteppe** [comment]
Oh man, the midpoint calculation was wrong! Not sure what I did there. Anyway, it should be better now. And I also noticed that it doesn't care about collision masks etc, so I made a quick fix for that. Try it now?

**Comment by jtenner** [comment]
I want to try it right away. Yet, I cannot get to the computer until tomorrow. 

Thank you for the fix, I will confirm in 18 hours or so.

**Comment by jtenner** [comment]
I replaced the source files in my project internally (because there is no build of it, I had to make it myself using webpack.)

After initial testing, and still having the problem, I cloned the repo and `npm install`ed the dependencies manually. After placing it in `node_modules` I found the problem to still exist.

I have made no changes to the example except that I am running the example locally.

![Problem](http://zippy.gfycat.com/SharpVibrantImperialeagle.gif)

- The bullet immediately overlaps with the player (despite the collision mask being set correctly)
```javascript
var CollisionGroups = {
  PLAYER: 1,
  TERRAIN: 2,
  BULLETS: 4
};
var playerShape = new p2.Circle({
  radius: playerRadius,
  collisionGroup: CollisionGroups.PLAYER,
  collisionMask: CollisionGroups.TERRAIN
});
var wallShape = new p2.Plane({
  collisionGroup: CollisionGroups.TERRAIN,
  collisionMask: CollisionGroups.PLAYER | CollisionGroups.BULLETS
});
```

- The bullet collides with the wall when it doesn't hit the player and doesn't move afterwards
(see above) only now sometimes the bullet actually sticks to the wall or dies out because of speed constraints.
`Math.hypot(bullet.velocty[0], bullet.velocity[1]) < maxBulletSpeed`

- After realizing the bullets wouldn't bounce, I set the defaultContactMaterial.restitution property
`world.defaultContactMaterial.restitution = 0.6;`
This resulted in every bullet sticking to the Plane. 

Thoughts? I want to help. Please let me know if I can.
-Josh

**Comment by jtenner** [comment]
@schteppe 

I got it!

The ccd algorithm sees the overlap with the plane every frame and the ccd algoritm runs... EVERY FRAME.  The bullet snaps to the wall with the correct velocity and never moves.

This is clearly not intended behavior.  

![image](https://cloud.githubusercontent.com/assets/3761339/11638751/889832b4-9cf6-11e5-8406-bf3d48b9fd00.png)

Checking the `firstImpact` property on the equation might be the way to solve the problem, but I do not know how physics engines work and I might be in over my head.

**Comment by schteppe** [comment]
It should only run once and move the bullet close to the wall with a small overlap. Next tick, the contact will change the velocity of the ball, making the CCD algo not run.

This is what I thought I fixed with the last update.. I even made a test locally. Can you reconfirm that you have the very latest commits? Thanks.

**Comment by jtenner** [comment]
![image](https://cloud.githubusercontent.com/assets/3761339/11639381/7c135e02-9cfa-11e5-84e3-1dee189fc53c.png)

Yes, it definitely is the correct version.

The ccd algorithm keeps running every frame. :(

**Comment by jtenner** [comment]
1. ccd is running for bullets vs players for no reason
2. ccd is running every frame for `Particle`s and stuff that continuously overlaps over objects.

**Comment by schteppe** [comment]
Hmm gotta test this some more, might get some time for that tomorrow.

**Comment by jtenner** [comment]
If I fire one bullet while stationary, sometimes the bullet stays stationary.  This is because collision detection is happening bullet vs player.  Upon further investigation, I discovered that the ray collision group and collision mask are always both -1.

![image](https://cloud.githubusercontent.com/assets/3761339/11640540/0ae73062-9d01-11e5-80fb-25fa5cf54d2c.png)

I hope this helps.
-Josh

**Comment by jtenner** [comment]
It looks like narrowphase is called to confirm that the bullet can overlap the player, and it yields nothing.  The collision detection on narrowphase works exactly as intended.  However, `body.integrateToTimeOfImpact` happens in the phase where the bullet moves freely.  This is probably a design flaw because it's already determined it won't collide with the player.

I hope this helps.

-Josh

**Comment by jtenner** [comment]
Line 163 of src/collision/Ray.js has a line that always yields false.

The condition looks like this:
```javascript
(this.collisionGroup & shape.collisionMask) === 0 || 
  (shape.collisionGroup & this.collisionMask) === 0
```

Given: `this.collisionGroup` and `this.collisionMask` is always `-1` when executing.

```javascript
> -1 & 2
//2
> -1 & 4
//4
```

Conclusion: this is probably where the problem becomes noticeable, but in reality, the `collisionGroup` and `collisionMask` properties are never set at all and have no bearing on the execution of intersectBody.

Mask information shouldn't be handled on the body level anyway.

This problem needs to be traced down to the Raycasting algorithm in each shape object.

I need more time to investigate, but I think it's safe to remove that offending line.

-Josh

**Comment by schteppe** [comment]
Made another quick fix. It runs a raycast for each shape now. Could probably be optimized, but whatever.

Also made it ignore the case when timeOfImpact is zero... I got that while shooting particles at circles in one of my demos, hopefully it works for you too now.

**Comment by jtenner** [comment]
![Problem](http://zippy.gfycat.com/EnlightenedDearestBee.gif)

I have so much to learn, but that worked perfectly!  Thank you so much!

**Comment by schteppe** [comment]
I'm glad to help :)

**Comment by jtenner** [comment]
Before this issue is closed, I've noticed another behavior that could potentially be problematic.

![Problem](http://zippy.gfycat.com/CanineFrequentAfricanelephant.gif)

It appears that bullets with ccd enabled can get stored inside of other objects. I'm not sure how to replicate it either, but here it is actually happening. 

In this case, the circles are point masses using the `Particle` class.

I know I've been knocking at this problem for days now, especially after many many changes. However, it might cause problems if bullets store up inside a box over time and randomly kill a player.

Any ideas?

**Comment by schteppe** [comment]
A never ending problem :)

Does it help to set skipBackfaces: true for the CCD ray?

Does it help to set contactSkinSize:0 for all involved shapes?

**Comment by jtenner** [comment]
Setting the contactSkinSize to 0 for the default material for testing purposes seems to really reduce the likelihood of the bullets getting stuck inside a box.

Behavior that is still occuring:
1. If the bullet spawns inside a box, it gets "stuck"
2. If the bullet as a result of collisions detection is forced inside a box, it gets "stuck"

How do I set "skipBackfaces" inside normal code?

**Comment by schteppe** [comment]
Interesting.. No you can probably not. You have to edit the p2 source for that... Or monkey patch the whole CCD method on p2.Body.prototype.

Can fix tomorrow probably. Not sure why I didn't skip backfaces in the first place.

**Comment by jtenner** [comment]
It should be possible to attach and modify the default ray properties for testing purposes inside the code.  Care to add them to the world object, or better yet, add them to the `Ray` function?  It might be useful for later.

**Comment by schteppe** [comment]
Yes, it is possible. But, I try to avoid putting properties on objects if it's not necessary. You see, property names cannot be minified, but variables in a closed scope can. It would make hacking easier though, yes.

**Comment by schteppe** [comment]
Fixed more stuff. f409f2f5d4a0aa8bd07785625dc8129797744e90

**Comment by schteppe** [comment]
Does it work better now? It should :)

**Comment by jtenner** [comment]
I'm sorry I couldn't get around to this yesterday. 

Confirming I have the latest version running...
![image](https://cloud.githubusercontent.com/assets/3761339/11720725/87f85390-9f2e-11e5-8a3a-78af2359fdb2.png)

Particles behave exactly as expected, but any shape that has a size and spawns before the front face will still get "stuck"

![Problem](http://zippy.gfycat.com/QuerulousBlushingFluke.gif)

As of right now, I feel for you, and so does picard.

![Problem](http://www.troll.me/images/facepalm-picard/my-god-this-will-never-end-thumb.jpg)

**Comment by schteppe** [comment]
But it gets sooo much more stable :+1: 

Actually, it should work even if you used ```disableBodyCollision```. But i fixed it anyway now. Quick and dirty. Optimize later etc. Update again :)

Just to reconfirm, do you have contactSkinSize=0 on all shapes?

**Comment by jtenner** [comment]
You bet!  Thank you so much.

**Comment by pkieltyka** [comment]
hey guys, thanks for the details in here. I'm running up against the same issue in a zero-gravity space shooter with two ships (dynamic) shooting at each other with bullets having varying velocity. In some case, especially when I fire a lot of bullets at the other ship they get stuck at the edges of the enemy ship. Any suggestions please? I am using the latest master of p2.js

**Comment by pkieltyka** [comment]
by the way @jtenner and @schteppe -- I was able to reproduce ccd stickiness issue in a simple modification to the ccd.html example.

Here is the fiddle demonstrating the bug: https://jsfiddle.net/ybL2ar0q/1/

If you run it, you will see the ball moves quickly towards to other circle, stops right in front of it, gets stuck then eventually after a few seconds continues and makes contact and bounces back. You can adjust the radius on line 28 to play with the behaviour. This is exactly what I was experiencing in my game as well. It seems if the sizes of the objects in the ccd differ, then this bug occurs. 

another question is, is it possible to make CCD only run when its in a certain proximity to an object? in my game, if I fired a lot of bullets with ccd then eventually the FPS would get quite slow, but without ccd I could shoot a ridiculously large amount of bullets and still have great FPS.

**Comment by jtenner** [comment]
Ccd stickiness is on the current release of p2.js, but is fixed on the main branch, at least it was for me. 

I remember helping @schteppe fix this problem

**Comment by pkieltyka** [comment]
@jtenner you're right, I rebuilt build/p2.js from master and ccd.html with the changes works correctly now. Unfortunately my game is still experiencing the stickiness so I'll try to reproduce the issue again. Happy New Year!

**Comment by jtenner** [comment]
@pkieltyka could you build a codepen with the problem?

**Comment by pkieltyka** [comment]
@jtenner sure thing. I'll prepare something in a few hours that reproduces what I'm experiencing

**Comment by pkieltyka** [comment]
@jtenner okay, false alarm. After trying to write an isolated example of the bug, I found out my problem was in fact how I was dealing with beginContact events.

it would be cool if I could set collisionResponse = false on a shape after beginContact to prevent the collision behaviour (I tried). In my case, I have bouncing bullets off walls and then exploding bullets immediately on contact with another ship (bullets just disappear and do not complete the collision), I will need to make two of the same shapes with different masks, one for walls and one for enemy ships.
